fs-index/src/index.rs

```diff
@@ -1,1120 +1,501 @@
+/// ```no_run
+/// use std::path::Path;
+/// use fs_index::{ResourceIndex, load_or_build_index};
+/// use dev_hash::Crc32;
+///
+/// // Define the root path
+/// let root_path = Path::new("animals");
+///
+/// // Build the index
+/// let index: ResourceIndex<Crc32> = ResourceIndex::build(root_path).expect("Failed to build index");
+/// // Store the index
+/// index.store().expect("Failed to store index");
+///
+/// // Load the stored index
+/// let mut loaded_index: ResourceIndex<Crc32> = load_or_build_index(root_path, false).expect("Failed to load index");
+///
+/// // Update the index
+/// loaded_index.update_all().expect("Failed to update index");
+///
+/// // Get a resource by path
+/// let _resource = loaded_index
+///     .get_resource_by_path("cat.txt")
+///     .expect("Resource not found");
+///
+/// // Track the removal of a file
+/// loaded_index
+///     .track_removal(Path::new("cat.txt"))
+///     .expect("Failed to track removal");
+///
+/// // Track the addition of a new file
+/// loaded_index
+///     .track_addition(Path::new("dog.txt"))
+///     .expect("Failed to track addition");
+///
+/// // Track the modification of a file
+/// loaded_index
+///     .track_modification(Path::new("dog.txt"))
+///     .expect("Failed to track modification");
+/// ```
+#[derive(Clone, Debug)]
+pub struct ResourceIndex<Id>
 where
-    Id: ResourceId,
+    Id: Eq + Hash,
 {
-    entries
-        .into_iter()
-        .filter_map(|(path_buf, entry)| {
-            let metadata = entry.metadata().ok()?;
-
-            let path = path_buf.as_canonical_path();
-            let result = scan_entry(path, metadata);
-            match result {
-                Err(msg) => {
-                    log::error!(
-                        "Couldn't retrieve metadata for {}:\n{}",
-                        path.display(),
-                        msg
-                    );
-                    None
-                }
-                Ok(entry) => Some((path_buf, entry)),
-            }
-        })
-        .collect()
+    /// The root path of the index (canonicalized)
+    pub(crate) root: PathBuf,
+    /// A map from resource IDs to resources
+    ///
+    /// Multiple resources can have the same ID (e.g., due to hash collisions
+    /// or files with the same content)
+    pub(crate) id_to_resources: HashMap<Id, Vec<IndexedResource<Id>>>,
+    /// A map from resource paths to resources
+    pub(crate) path_to_resource: HashMap<PathBuf, IndexedResource<Id>>,
 }
 
-fn is_hidden(entry: &DirEntry) -> bool {
-    entry
-        .file_name()
-        .to_str()
-        .map(|s| s.starts_with('.'))
-        .unwrap_or(false)
+/// Represents the result of an update operation on the ResourceIndex
+#[derive(PartialEq, Debug)]
+pub struct IndexUpdate<Id: ResourceId> {
+    /// Resources that were added during the update
+    added: Vec<IndexedResource<Id>>,
+    /// Resources that were modified during the update
+    modified: Vec<IndexedResource<Id>>,
```

Comment:
Am I right that modification is an update of content by some path?

If a resource changes its location, is it represented as removal and addition at the same time?

Reply:
The logic for tracking modifications is to first track removal and then track addition. This is to make sure that the ID and last modified time are updated.

> If a resource changes its location, is it represented as removal and addition at the same time?

Yes, if a resource changes its path, it is treated as an addition and removal. This is because we use paths as the unique identifier for file systems. Hashes, including cryptographic hashes, can be the same if the content is identical.